lib/ansible/modules/network/ftd/ftd_install.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/python
+
+# Copyright (c) 2019 Cisco and/or its affiliates.
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = """
+---
+module: ftd_install
+short_description: Installs FTD pkg image on the firewall
+description:
+  - Provisioning module for FTD devices that installs ROMMON image (if needed) and
+    FTD pkg image on the firewall.
+  - Can be used with `httpapi` and `local` connection types. The `httpapi` is preferred,
```

Comment:
@justjais Is `httpapi` the recommended method for requests instead of local in the future?

Reply:
@justjais, is `delegate_to: localhost` supposed to be its replacement when `local` connection is deprecated (for example, when API is not accessible and `httpapi` connection cannot be used because of this)?

Reply:
@kbreit yes, `httpapi` connection is preferred over `local`, n that's why in network platform other connection plugins like: `network_cli/netconf` along with `httpapi` were introduced.

@annikulin `delegate_to: localhost` and `connection: local` with respect to network means the same, also if I understand your example with respect to `API` not accessible correctly then u meant was if the server is inaccessible and if that's the case then even `local` connection won't work coz all these connection plugins runs the module locally only.

Reply:
@justjais, yes, and in case the `API` is not accessible, the playbook execution will fail even before executing the first task (when establishing the connection). In this case, we would still like to connect to the device and execute commands via console server. 

What type of connection is recommended in this case? Should it be still `httpapi` with `delegate_to: localhost` option?

Reply:
To add to @justjais's comment, we plan to deprecate `connection=local` for network platforms that are supported by Ansible Network team (ios, eos, junos, iosxr, nxos, vyos) in favor of first-class network connection plugins (network_cli/httpapi/netconf). 

For new network platform support added by the community, it is recommended to use first class network connection plugin wherever applicable.

If the connection to console happens over ssh you can use `connection=network_cli` in that case. This connection creates the pseudo terminal over ssh to talk to remote host Refer [network_cli](https://github.com/ansible/community/blob/master/group-network/network_dev_network_cli.rst)

For all the network connection plugins the module runs locally on control node itself, hence `delegate_to: localhost` with httpapi connection won't have any effect (unless jumphost is used)

Reply:
@ganeshrn, @justjais, thanks for your comments. Consolidating network modules and making them support network connection plugins only sounds reasonable.

As of now, we've used the [`unicon`](https://pypi.org/project/unicon/) library developed by Cisco to connect to the device over SSH and execute commands. Having taken a brief look at `network_cli` connection, the ideas look similar but the approach is a bit different: `unicon` has a concept of dialogs defining a sequence of commands you want to execute, a sequence of outputs you expect to receive and behavior of what to do when the command fails or times out. I'm pretty sure we can implement similar functionality with the help of `network_cli`, but it requires thorough investigation and planning together with you guys as owners of this plugin.

Considering that this module is already useful for FTD customers, would it make sense to publish this module as is (with dependency on `unicon` library and support for `local` connection type) as a short term solution and plan how to replace it with `network_cli` connection as a long-term plan and do it as a part of the next release?

Reply:
>  would it make sense to publish this module as is (with dependency on unicon library and support for local connection type) as a short term solution and plan how to replace it with network_cli connection as a long-term plan and do it as a part of the next release?

Yes, that should work IMO. FYI Ansible follows 4 version deprecation cycle that is if the code (options) available in a stable release is deprecated it can be removed after four releases, typical Ansible release cycle is around 4 to 6 months.

Reply:
Got it, thanks @ganeshrn.

@justjais, would it work for you? So that we leave the `local` connection in this PR and come up with a plan how to get rid of it down the road.

Reply:
@annikulin yes, I agree with @ganeshrn comments. I'll approve and merge the PR.

Reply:
Thank you, @justjais.